design/design-doc-rbac-manager.md

```diff
@@ -0,0 +1,777 @@
+
+Note that these role names are derived from the active `ProviderRevision`. In
+this example `7f63e3661` is an abbreviated hash. The `aggregate-to-edit` role
+aggregates to the `crossplane`, `crossplane-admin` and `crossplane-edit` roles,
+while the `aggregate-to-view` role aggregates to the `crossplane-view` role. The
+`system` role does not aggregate; it is granted by the RBAC manager to the
+service account under which the provider pod runs.
+
+To create a cluster role for a `Provider` (package), the RBAC manager:
+
+1. Finds its active `ProviderRevision`.
+1. Lists all CRDs that are controlled by that `ProviderRevision`.
+1. Creates a `ClusterRole` that grants access to the defined CRs.
+
+In the case of the `system` role the RBAC manager creates a `ClusterRoleBinding`
+to the `ServiceAccount` under which the provider runs. This service account is
+also controlled by the aforementioned `ProviderRevision`, and can be discovered
+similarly.
+
+Below is an example of the `system` provider role and its role binding.
+
+```yaml
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: crossplane:provider:7f63e3661:system
+rules:
+- apiGroups: [provider.example.org]
+  resources:
+  - examplemanageds
+  - examplemanageds/status
+  - exampleproviderconfigs
+  - exampleproviderconfigs/status
+  verbs: [get, list, watch, update, patch]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: crossplane:provider:7f63e3661:system
+roleRef:
+  apiVersion: rbac.authorization.k8s.io/v1
+  kind: ClusterRole
+  name: crossplane:provider:7f63e3661:system
+subjects:
+- kind: ServiceAccount
+  name: crossplane:provider:7f63e3661
+  namespace: crossplane-system
+```
+
+Below is an example of an `edit` provider role. Note that in practice the role
+would likely have many more rules - one for each custom resource defined by the
+provider.
+
+```yaml
+---
+# This role aggregates to the crossplane and crossplane-edit roles. It thus
+# aggregates transitively to the crossplane-admin role via crossplane-edit.
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: crossplane:provider:7f63e3661:aggregate-to-edit
+  labels:
+    rbac.crossplane.io/aggregate-to-crossplane: "true"
+    rbac.crossplane.io/aggregate-to-edit: "true"
+rules:
+- apiGroups: [provider.example.org]
+  resources:
+  - examplemanageds
+  - exampleproviderconfigs
+  verbs: [*]
+```
+
+### Opting out of RBAC Management
+
+Building and deploying the RBAC manager as a distinct process allows Crossplane
+users to completely opt-out of RBAC management. Users may choose not to deploy
+the RBAC manager without compromising any core Crossplane functionality. In
+addition to completely disabling the RBAC manager it will be possible to limit
+the scope of the RBAC manager via a flag. The following levels will exist:
+
+* `--manage=all` - Include all functionality described in this design.
```

Comment:
```suggestion
* `--manage=all` - Include all functionality described in this design. This is the default.
```

`--manage=all` will be the default, right?

Reply:
Right.